lizard_languages/clike.py

```diff
@@ -72,16 +72,85 @@ def _typedef(self, _, tokens):
         self.context.add_condition(-tokens.count("&&"))
 
 
+# pylint: disable=R0903
 class CLikeNestingStackStates(CodeStateMachine):
-    # pylint: disable=R0903
+    """Machinery to track nesting levels of tokens.
+
+    The main indicators of nesting are braces: '{' and '}'.
+    However, the code is complicated due to braceless structures,
+    which are control-flow structures in C-like languages.
+    Moreover, using braces as nesting level indicators leads to a big caveat;
+    C++ initializer lists, uniform initialization,
+    and any other constructs with braces add extra nesting level.
+
+    Another complication comes from nested classes inside function bodies
+    and control-flow declarations/bodies.
+    The handling of these complex cases is unspecified and can be ignored.
+    """
+
+    # Beasts that can be defined within one line without braces.
+    __braceless_structures = set(['if', 'else', 'for', 'while', 'do',
+                                  'switch'])
+    __paren_count = 0  # Used only to tackle the beasts.
+    __braceless = None  # Applies only to the beasts.
+    __structure_brace_stack = []  # Boolean stack for structures' brace states.
+
+    def __pop_braceless_structures(self):
+        """Pops structures up to the one with braces."""
+        self.context.pop_nesting()
+        is_structure = None
+        if self.__structure_brace_stack:
+            is_structure = self.__structure_brace_stack.pop()
+
+        while (is_structure is not None and self.__structure_brace_stack and
+                self.__structure_brace_stack[-1]):
+            self.__structure_brace_stack.pop()
+            self.context.pop_nesting()
+
+    def __else_if_structure(self, token):
+        """Handles possible compound 'else if' after 'else' token."""
+        self._state = self.__declare_structure
+        if token != "if":
+            self._state(token)
+
+    def __declare_structure(self, token):
+        """Ignores structures between parentheses on structure declaration."""
+        if token == "(":
+            self.__paren_count += 1
+        elif token == ")":
+            assert self.__paren_count > 0
```

Comment:
But some of them installed lizard using pip, and the users don’t event know it’s Python behind it.

> On 2 Jun 2016, at 2:22 AM, Olzhas Rakhimov <notifications@github.com> wrote:
> 
> In lizard_languages/clike.py <https://github.com/terryyin/lizard/pull/123#discussion_r65415551>:
> 
> > +                self.__structure_brace_stack[-1]):
> > +            self.__structure_brace_stack.pop()
> > +            self.context.pop_nesting()
> > +
> > +    def __else_if_structure(self, token):
> > +        """Handles possible compound 'else if' after 'else' token."""
> > +        self._state = self.__declare_structure
> > +        if token != "if":
> > +            self._state(token)
> > +
> > +    def __declare_structure(self, token):
> > +        """Ignores structures between parentheses on structure declaration."""
> > +        if token == "(":
> > +            self.__paren_count += 1
> > +        elif token == ")":
> > +            assert self.__paren_count > 0
> You don't have to remove them.
> C'mon, don't they know how to suppress assertion?
> Let them run Python with optimization flag, i.e. -O.
> This will suppress asserts for them.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/terryyin/lizard/pull/123/files/47af1f2b3bb39f4719968a72dc38a4f41f1995a3#r65415551>, or mute the thread <https://github.com/notifications/unsubscribe/AAwJYg3R22k4B8Ku7uo-A4ldpSKwI6Jxks5qHc3vgaJpZM4Id8An>.
>